sdk/helper/keysutil/policy.go

```diff
@@ -1973,3 +1969,153 @@ func (p *Policy) EncryptWithFactory(ver int, context []byte, nonce []byte, value
 
 	return encoded, nil
 }
+
+func (p *Policy) UpdateKeyVersion(ctx context.Context, storage logical.Storage, key []byte, isPrivateKey bool, keyVersion int) error {
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return err
+	}
+
+	// Validations
+	if !p.Type.ImportPublicKeySupported() {
+		return errors.New("provided type does not support importing key versions")
+	}
+
+	publicKeyImported := keyEntry.IsPublicKeyImported()
+	if publicKeyImported && !isPrivateKey {
+		return errors.New("cannot add a public key to a key version that already has a public key set")
+	}
+
+	if !publicKeyImported {
+		return errors.New("private key imported, key version cannot be updated")
+	}
+
+	// Parse key
+	parsedPrivateKey, err := x509.ParsePKCS8PrivateKey(key)
+	if err != nil {
+		return fmt.Errorf("error parsing asymmetric key: %s", err)
+	}
+
+	switch parsedPrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		ecdsaKey := parsedPrivateKey.(*ecdsa.PrivateKey)
+		pemBlock, _ := pem.Decode([]byte(keyEntry.FormattedPublicKey))
+		publicKey, err := x509.ParsePKIXPublicKey(pemBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse key entry public key: %v", err)
+		}
+		if !publicKey.(*ecdsa.PublicKey).Equal(ecdsaKey.PublicKey) {
+			return fmt.Errorf("cannot import key, key pair does not match")
+		}
+	case *rsa.PrivateKey:
+		rsaKey := parsedPrivateKey.(*rsa.PrivateKey)
+		if !rsaKey.PublicKey.Equal(keyEntry.RSAPublicKey) {
+			return fmt.Errorf("cannot import key, key pair does not match")
+		}
+	}
+
+	err = keyEntry.parseFromKey(p.Type, parsedPrivateKey, isPrivateKey)
+	if err != nil {
+		return err
+	}
+
+	p.Keys[strconv.Itoa(keyVersion)] = keyEntry
+
+	return p.Persist(ctx, storage)
+}
+
+func (ke *KeyEntry) parseFromKey(PolKeyType KeyType, parsedKey any, isPrivateKey bool) error {
```

Comment:
I'm fine deleting the public key; we already guaranteed they'd match on import I believe.

Reply:
I see your point, `RSAPublicKey` would have a value assigned in versions where a "public key" was imported first otherwise would be empty. Assuming what Alex said and that there isn't any use for it once the private counterpart is imported I also agree we should "delete" it.

Reply:
My guess is perhaps the authors thought there could be multiple formats for RSA keys (PKCS#1 vs PKCS#8) and wanted the users to choose between them, so they didn't marshal in either format to just return it? :shrug: